src/Optimizer/Statistics/DeriveStatistics.cpp

```diff
@@ -0,0 +1,496 @@
+                statistics.addColumnStatistics(column, input->getColumnStatistics(column)->clone());
+    };
+
+    /// 2. calculate for prewhere filters
+    if (step.getPrewhereInfo())
+    {
+        const auto & prewhere_info = step.getPrewhereInfo();
+        if (prewhere_info->row_level_filter)
+        {
+            statistics = PredicateStatsCalculator::calculateStatistics(
+                prewhere_info->row_level_filter, prewhere_info->row_level_column_name, statistics);
+
+            statistics.removeColumnStatistics(prewhere_info->row_level_column_name);
+            append_column_stats();
+        }
+
+        if (prewhere_info->prewhere_actions)
+        {
+            statistics = PredicateStatsCalculator::calculateStatistics(
+                prewhere_info->prewhere_actions, prewhere_info->prewhere_column_name, statistics);
+
+            if (prewhere_info->remove_prewhere_column)
+                statistics.removeColumnStatistics(prewhere_info->prewhere_column_name);
+            append_column_stats();
+        }
+    }
+
+    /// 3. calculate for pushed down filters
+    for (size_t i = 0; i < step.getFilters().size(); i++)
+    {
+        const auto & predicate = step.getFilters()[i];
+        const auto & predicate_node_name = step.getFilterNodes().nodes[i]->result_name;
+        statistics = PredicateStatsCalculator::calculateStatistics(predicate, predicate_node_name, statistics);
+        append_column_stats();
+    }
+
+    /// Remove the additional columns and add missing ones.
+    adjustStatisticsByColumns(statistics, output_columns);
+
+    statistics.adjustStatistics();
+    return statistics;
+}
+
+Stats DeriveStatistics::visit(ExpressionStep & step)
+{
+    chassert(input_statistics.size() == 1);
+    const auto & action_dag = step.getExpression();
+    Stats statistics = ExpressionStatsCalculator::calculateStatistics(action_dag, input_statistics.front());
+    return statistics;
+}
+
+Stats DeriveStatistics::visit(FilterStep & step)
+{
+    chassert(input_statistics.size() == 1);
+    Stats statistics
+        = PredicateStatsCalculator::calculateStatistics(step.getExpression(), step.getFilterColumnName(), input_statistics.front());
+    return statistics;
+}
+
+Stats DeriveStatistics::visit(AggregatingStep & step)
+{
+    Stats statistics;
+
+    Names input_names = step.getInputStreams().front().header.getNames();
+    Names output_names = step.getOutputStream().header.getNames();
+
+    const auto & input = input_statistics.front();
+
+    /// 1. initialize statistics
+    if (step.isGroupingSets())
+        statistics.addColumnStatistics(output_names[0], ColumnStatistics::unknown());
+
+    for (size_t i = 0; i < input_names.size(); i++)
+    {
+        /// The input name and output name are one-to-one correspondences.
+        auto column_stats = input.getColumnStatistics(input_names[i]);
+        statistics.addColumnStatistics(output_names[step.isGroupingSets() ? i + 1 : i], column_stats->clone());
```

Comment:
failing here on distributed join queries with seg fault for array indexing

/contrib/llvm-project/libcxx/include/vector:1456: std::vector<String, std::allocator<String>>::operator[][abi:v15000](unsigned long) @ 0x000000000a812442
build/../src/Optimizer/Statistics/DeriveStatistics.cpp:177: DB::DeriveStatistics::visit(DB::AggregatingStep&) @ 0x000000001ba61ae0

Reply:
Could you show the SQL?

Reply:
i have 1 cluster(removed default, substituted with a new cluster named cluster1) with 1 shard, 2 replicas and 2 distributed table on the same database on this cluster. 

query:
select count(*), hostName() as h from table1 as t1 join table2 as t2 on t1.val + 5 = t2.val group by h settings allow_experimental_parallel_reading_from_replicas = 0, allow_experimental_query_coordination = 1